File: service-maker/sources/apps/deepstream_test5_app/deepstream_test5.cpp
#include <string.h>

#include <fstream>
#include <iostream>
#include <memory>
#include <string>

#include "common_factory.hpp"
#include "lib/model_engine_watch_otf_trigger.hpp"
#include "lib/perf_monitor.hpp"
#include "pipeline.hpp"
#include "source_config.hpp"
#include "tiler_event_handler.hpp"

using namespace deepstream;

class MsgMetaGenerator : public BufferProbe::IBatchMetadataOperator {
public:
    MsgMetaGenerator(const std::map<uint32_t, SensorInfo> &sensor_map,
                     const std::vector<std::string> labels,
                     int frame_interval = 1)
        : sensor_map_(sensor_map), labels_(labels), frame_interval_(frame_interval)
    {
    }
    virtual ~MsgMetaGenerator() {}

    virtual probeReturn handleData(BufferProbe &probe, BatchMetadata &data)
    {
        FrameMetadata::Iterator frame_itr;
        for (data.initiateIterator(frame_itr); !frame_itr->done(); frame_itr->next()) {
            ObjectMetadata::Iterator obj_itr;
            for ((*frame_itr)->initiateIterator(obj_itr); !obj_itr->done(); obj_itr->next()) {
                if (0 == (frames_ % frame_interval_)) {
                    EventMessageUserMetadata event_user_meta;
                    if (data.acquire(event_user_meta)) {
                        auto source_id = (*frame_itr)->sourceId();
                        auto itr = sensor_map_.find(source_id);
                        if (itr != sensor_map_.end()) {
                            const std::string sensor = itr->second.sensor_id;
                            const std::string uri = itr->second.uri;
                            event_user_meta.generate(**obj_itr, **frame_itr, sensor, uri, labels_);
                            (*frame_itr)->append(event_user_meta);
                        } else {
                            std::cout << "Warning: sensor with source id " << source_id
                                      << " doesn't exist" << std::endl;
                        }
                    }
                }
            }
            frames_++;
        }

        return probeReturn::Probe_Ok;
    }

protected:
    int frames_ = 0;
    const std::map<uint32_t, SensorInfo> &sensor_map_;
    std::vector<std::string> labels_;
    int frame_interval_;
};

static void usage()
{
    std::cout << "Usage: deepstream-test5-app -s source-lists -c config-files [-l label-file] "
                 "[--perf-measurement-interval-sec] [--hide-stream-name]"
              << std::endl;
}

static std::string parse_config(std::vector<std::string> &arguments, const std::string flag)
{
    std::string str;
    auto itr = std::find(arguments.begin(), arguments.end(), flag);
    if (itr != arguments.end()) {
        auto next_itr = itr + 1;
        if (next_itr != arguments.end()) {
            str = *next_itr;
        }
    }
    return str;
}

static std::vector<std::string> split(const std::string &input, char delimiter)
{
    std::vector<std::string> result;
    std::istringstream stream(input);
    std::string token;

    while (std::getline(stream, token, delimiter)) {
        result.push_back(token);
    }

    return result;
}

static std::vector<std::string> parse_labels(const std::string &filename)
{
    std::ifstream file(filename);   // Open the file
    std::vector<std::string> lines; // Vector to store lines

    if (file.is_open()) { // Check if the file is open
        std::string line;
        while (std::getline(file, line)) { // Read each line
            std::string result = line;
            std::transform(line.begin(), line.end(), result.begin(),
                           [](unsigned char c) { return std::tolower(c); });
            lines.push_back(result); // Store the line in the vector
        }
        file.close(); // Close the file
    } else {
        std::cerr << "Unable to open file: " << filename << std::endl;
    }

    return lines;
}

int main(int argc, char *argv[])
{
    if (argc < 5) {
        usage();
        return 0;
    }

    std::string model_engine_file;
    uint32_t perf_interval = 1;
    std::vector<std::string> arguments;
    for (int n = 1; n < argc; n++) {
        arguments.push_back(argv[n]);
    }
    // find the config files
    std::vector<std::string> config_files = split(parse_config(arguments, "-c"), ',');
    std::vector<std::string> source_config_files = split(parse_config(arguments, "-s"), ',');
    std::string label_file = parse_config(arguments, "-l");
    std::string perf_interval_sec = parse_config(arguments, "--perf-measurement-interval-sec");
    bool hide_stream_name =
        std::find(arguments.begin(), arguments.end(), "--hide-stream-name") != arguments.end();
    if (config_files.empty() || source_config_files.empty()) {
        usage();
        return 0;
    }
    if (config_files.size() != source_config_files.size()) {
        std::cout << "The number of pipeline configuration file must be equal to that of the "
                     "source list file"
                  << std::endl;
        return 0;
    }

    std::vector<std::map<uint32_t, SensorInfo>> sensor_maps(config_files.size());
    std::vector<std::string> labels;
    if (!label_file.empty()) {
        labels = parse_labels(label_file);
    }

    try {
        std::vector<std::unique_ptr<Pipeline>> pipelines;
        std::vector<std::unique_ptr<NvDsModelEngineWatchOTFTrigger>> otf_triggers;
        std::vector<std::unique_ptr<NvDsTilerEventHandler>> tiler_event_handlers;
        std::vector<std::unique_ptr<PerfMonitor>> perf_monitors;
        for (unsigned int idx = 0; idx < config_files.size(); idx++) {
            std::map<uint32_t, SensorInfo> &sensor_map = sensor_maps[idx];
            SourceConfig source_config(source_config_files[idx]);
            std::string pipeline_name = "deepstream-test5-";
            pipeline_name += std::to_string(idx);
            pipelines.push_back(
                std::make_unique<Pipeline>(pipeline_name.c_str(), config_files[idx]));
            Pipeline &pipeline = *pipelines.back().get();
            std::string kafka_proto_lib_path;
            std::string msgbroker_conn_str;
            std::string msgconv_config_path;
            std::string msgbroker_config_path;
            uint32_t tiler_width = 0;
            uint32_t tiler_height = 0;
            std::string pgie = "pgie";

            pipeline["msgbroker"].getProperty("config", msgbroker_config_path, "proto-lib",
                                              kafka_proto_lib_path, "conn-str", msgbroker_conn_str);
            pipeline["msgconv"].getProperty("config", msgconv_config_path);
            pipeline["tiler"].getProperty("width", tiler_width, "height", tiler_height);
            pipeline[pgie].getProperty("model-engine-file", model_engine_file);

            // create the smart recording action
            auto object =
                CommonFactory::getInstance().createObject("smart_recording_action", "sr_action");
            auto *sr_action = dynamic_cast<SignalEmitter *>(object.get());
            if (!sr_action) {
                std::cerr << "Failed to create signal emitter" << std::endl;
                return -1;
            }

            sr_action->set("proto-lib", kafka_proto_lib_path, "conn-str", msgbroker_conn_str,
                           "msgconv-config-file", msgconv_config_path, "proto-config-file",
                           msgbroker_config_path, "topic-list", "test5-sr");

            std::string checkpoint = "checkpoint";
            if (source_config.useMultiUriSrcBin()) {
                // nvmultiurisrcbin used
                pipeline.add("nvmultiurisrcbin", "multiurisrcbin");
                pipeline["multiurisrcbin"]
                    .set(source_config.getProperties())
                    .set("uri-list", source_config.listUris())
                    .set("sensor-id-list", source_config.listSensorIds())
                    .set("sensor-name-list", source_config.listSensorNames());

                pipeline.link("multiurisrcbin", pgie);
            } else {
                // nvurisrcbin used
                auto &properties = source_config.getProperties();
                pipeline.add("nvstreammux", "mux", "batch-size", source_config.nSources(), "width",
                             tiler_width, "height", tiler_height);
                if (properties["gpu-id"]) {
                    pipeline["mux"].set("gpu-id", properties["gpu-id"].as<int>());
                }
                // create and add all the required elements
                for (size_t i = 0; i < source_config.nSources(); i++) {
                    // multiple sources added
                    std::string src_name = "src_";
                    src_name += std::to_string(i);
                    std::string uri = source_config.getSensorInfo(i).uri;
                    pipeline.add("nvurisrcbin", src_name, "uri", uri.c_str())
                        .link({src_name, "mux"}, {"vsrc_%u", ""});
                    pipeline[src_name].set(properties);
                    sensor_map[i] = SensorInfo{uri, std::to_string(i), src_name};
                }
                pipeline.link("mux", pgie);
            }
            pipeline.attach(checkpoint, new BufferProbe("message_generator",
                                                        new MsgMetaGenerator(sensor_map, labels)));

            // create model update watcher
            otf_triggers.push_back(std::make_unique<NvDsModelEngineWatchOTFTrigger>(
                &pipeline[pgie], model_engine_file));
            NvDsModelEngineWatchOTFTrigger &otf_trigger = *otf_triggers.back().get();

            // create the event monitor for the tiler
            tiler_event_handlers.push_back(std::make_unique<NvDsTilerEventHandler>(
                &pipeline["tiler"], &pipeline["osd"], &pipeline["sink"]));
            NvDsTilerEventHandler &tiler_event_handler = *tiler_event_handlers.back().get();

            if (!source_config.useMultiUriSrcBin()) {
                // add the smart recording controll to all the sources
                for (size_t i = 0; i < source_config.nSources(); i++) {
                    std::string src_name = "src_";
                    src_name += std::to_string(i);
                    sr_action->attach("start-sr", pipeline[src_name]);
                    sr_action->attach("stop-sr", pipeline[src_name]);
                    pipeline[src_name].connectSignal("smart_recording_signal", "sr", "sr-done");
                }
            }

            // performance monitor
            if (!perf_interval_sec.empty()) {
                perf_interval = std::stoi(perf_interval_sec);
            }

            perf_monitors.push_back(std::make_unique<PerfMonitor>(
                source_config.nSources(), perf_interval,
                source_config.useMultiUriSrcBin() ? "nvmultiurisrcbin" : "nvurisrcbin",
                !hide_stream_name));
            PerfMonitor &perf_monitor = *perf_monitors.back().get();
            perf_monitor.apply(pipeline["tiler"], "sink");

            // install keyboard listener
            pipeline.install([&perf_monitor](Pipeline &pipeline, int key) {
                switch (key) {
                case 'p':
                    pipeline.pause();
                    perf_monitor.pause();
                    break;
                case 'r':
                    pipeline.resume();
                    perf_monitor.resume();
                    break;
                case 'q':
                    pipeline.stop();
                    break;
                default:
                    break;
                }
            });

            // start pipeline with message monitor
            pipeline.start([&otf_trigger, &tiler_event_handler, &sensor_map, &perf_monitor](
                               Pipeline &p, const Pipeline::Message &msg) {
                const Pipeline::StateTransitionMessage *state_change_msg =
                    dynamic_cast<const Pipeline::StateTransitionMessage *>(&msg);
                const Pipeline::DynamicSourceMessage *source_change_msg =
                    dynamic_cast<const Pipeline::DynamicSourceMessage *>(&msg);
                // handle state change message
                if (state_change_msg) {
                    static bool tiler_event_handler_started = false;
                    if (!tiler_event_handler_started) {
                        Pipeline::State old_state = Pipeline::State::EMPTY;
                        Pipeline::State new_state = Pipeline::State::EMPTY;
                        state_change_msg->getState(old_state, new_state);
                        if (new_state == Pipeline::State::PLAYING &&
                            state_change_msg->getName() == "sink") {
                            tiler_event_handler.start();
                            tiler_event_handler_started = true;
                        }
                    }
                    static bool otf_trigger_started = false;
                    if (!otf_trigger_started) {
                        otf_trigger.start();
                        otf_trigger_started = true;
                    }
                }

                // handle source add/remove message
                if (source_change_msg) {
                    auto source_id = source_change_msg->getSourceId();
                    if (source_change_msg->isSourceAdded()) {
                        auto uri = source_change_msg->getUri();
                        auto sensor_id = source_change_msg->getSensorId();
                        auto sensor_name = source_change_msg->getSensorName();
                        if (sensor_map.find(source_id) == sensor_map.end()) {
                            sensor_map[source_id] = SensorInfo{uri, sensor_id, sensor_name};
                        } else {
                            std::cout << "Warning, sensor with source id " << source_id
                                      << " already added" << std::endl;
                        }
                        perf_monitor.addStream(source_id, uri.c_str(), sensor_id.c_str(),
                                               sensor_name.c_str());
                    } else {
                        sensor_map.erase(source_id);
                        perf_monitor.removeStream(source_id);
                    }
                }
            });
        }
        // wait for all the pipelines to end
        for (auto &pipeline : pipelines) {
            pipeline->wait();
        }
    } catch (const std::exception &e) {
        std::cerr << e.what() << std::endl;
        return -1;
    }

    return 0;
}

File: service-maker/sources/apps/python/pipeline_api/deepstream_test5_app/deepstream_test5.py
#
# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#
# NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
# property and proprietary rights in and to this material, related
# documentation and any modifications thereto. Any use, reproduction,
# disclosure or distribution of this material and related documentation
# without an express license agreement from NVIDIA CORPORATION or
# its affiliates is strictly prohibited.
#

from pyservicemaker import Pipeline, PipelineState, StateTransitionMessage, DynamicSourceMessage, Probe, BatchMetadataOperator
from pyservicemaker import osd, signal, utils
from pyservicemaker import CommonFactory, SourceConfig, SensorInfo
from multiprocessing import Process
import sys
import argparse
import platform

PGIE = "pgie"
CHECKPOINT = "queue_tee"
TILER = "tiler"
MSGBROKER = "msgbroker"
MSGCONV = "msgconv"
OSD = "osd"
SINK = "sink"

class EventMessageGenerator(BatchMetadataOperator):
    """
    generate event message user metadata for downstream nvmsgconv
    to produce message payloads, which will be posted to the remote
    server by msgboker, e.g kafka
    """
    def __init__(self, sensor_map, labels):
        super().__init__()
        self._sensor_map = sensor_map
        self._labels = labels

    def handle_metadata(self,  batch_meta, frame_interval=1):
        for frame_meta in batch_meta.frame_items:
            frame_num = frame_meta.frame_number
            for object_meta in frame_meta.object_items:
                if not (frame_num % frame_interval):
                    event_msg = batch_meta.acquire_event_message_meta()
                    if event_msg:
                        source_id = frame_meta.source_id
                        sensor_info = self._sensor_map[source_id] if source_id in self._sensor_map else None
                        if sensor_info != None:
                            sensor_id = sensor_info.sensor_id if sensor_info else "N/A"
                            uri = sensor_info.uri if sensor_info else "N/A"
                            event_msg.generate(object_meta, frame_meta, sensor_id, uri, self._labels)
                            frame_meta.append(event_msg)

def on_message(message, osd_event_handler, sensor_map, perf_monitor, model_engine_monitor):
    if isinstance(message, StateTransitionMessage):
        if message.new_state == PipelineState.PLAYING and message.origin == "sink":
            if not osd_event_handler.started:
                osd_event_handler.start()
            if not model_engine_monitor.started:
                model_engine_monitor.start()
    elif isinstance(message, DynamicSourceMessage):
        source_id = message.source_id
        if message.source_added:
            if source_id in sensor_map:
                print(f"Warning, sensor with source id {source_id} already added")
            else:
                sensor_map[source_id] = SensorInfo(
                    sensor_id=message.sensor_id,
                    sensor_name=message.sensor_name,
                    uri=message.uri)
            perf_monitor.add_stream(
                source_id=source_id,
                sensor_id=message.sensor_id,
                sensor_name=message.sensor_name,
                uri=message.uri)
        else:
            del sensor_map[source_id]
            perf_monitor.remove_stream(source_id)

def main(args):
    # Check input arguments
    if len(args.pipeline_config) != len(args.source_config):
        sys.stderr.write("Number of pipeline config must equal to that of source config")
        sys.exit(1)
    labels = []
    if args.label_file:
        with open(args.label_file, 'r') as f:
            for label in f:
                labels.append(label.strip().lower())

    source_maps = [dict() for _ in args.pipeline_config]
    reference_holders = []
    pipelines = []
    pipeline_id = 0
    for source_config_file, pipeline_config_file in zip(args.source_config, args.pipeline_config):
        sensor_map = source_maps[pipeline_id]
        pipeline_name = f"deepstream-test5-{pipeline_id}"
        pipeline = Pipeline(name=pipeline_name, config_file=pipeline_config_file)
        pipelines.append(pipeline)
        source_config = SourceConfig()
        source_config.load(source_config_file)
        source_properties = dict(source_config.source_properties)
        use_multisrcbin = True if source_config.source_type == "nvmultiurisrcbin" else False
        use_camerabin = True if source_config.source_type == "camerabin" else False
        if use_multisrcbin:
            source_properties["uri-list"] = ','.join([source.uri for source in source_config.sensor_list])
            source_properties["sensor-id-list"] = ','.join([source.sensor_id for source in source_config.sensor_list])
            source_properties["sensor-name-list"] = ','.join([source.sensor_name for source in source_config.sensor_list])
            pipeline.add("nvmultiurisrcbin", "source", source_properties).link("source", PGIE)
        elif use_camerabin:
            streammux_properties = {
                "batch-size": len(source_config.camera_list),
                "width": pipeline[TILER].get("width"),
                "height": pipeline[TILER].get("height"),
                "live-source": 1,
                "batched-push-timeout": 40000
            }
            pipeline.add("nvstreammux", "mux", streammux_properties)
            if "gpu-id" in source_config.source_properties:
                pipeline["mux"].set("gpu-id", source_config.source_properties["gpu-id"])
            for i, camera_info in enumerate(source_config.camera_list):

                invalid_value_check = ""
                if not camera_info.camera_type:
                    invalid_value_check = "camera-type"
                if not camera_info.camera_width:
                    invalid_value_check = "camera-width"
                if not camera_info.camera_height:
                    invalid_value_check = "camera-height"
                if not camera_info.camera_fps_n:
                    invalid_value_check = "camera-fps-n"
                if not camera_info.camera_fps_d:
                    invalid_value_check = "camera-fps-d"

                if len(invalid_value_check):
                    print(f"No Value for {invalid_value_check}")
                    sys.exit()

                src_name = f"src_{i}"
                src_cap_filter = f"{src_name}_cap_filter"
                capfilterCaps = f"video/x-raw(memory:NVMM), format={camera_info.camera_video_format if camera_info.camera_video_format else 'NV12'}, width={camera_info.camera_width}, height={camera_info.camera_height}, framerate={camera_info.camera_fps_n}/{camera_info.camera_fps_d}"
                pipeline.add("capsfilter", src_cap_filter, {"caps": capfilterCaps})
                src_queue_mux = f"{src_name}_queue_mux"
                pipeline.add("queue", src_queue_mux)

                if camera_info.camera_type == "CSI":
                    if (camera_info.camera_csi_sensor_id == None):
                        print("camera-csi-sensor-id required")
                        sys.exit()
                    else:
                        pipeline.add("nvarguscamerasrc" if platform.processor() == "aarch64" else "videotestsrc", src_name, {"sensor-id": camera_info.camera_csi_sensor_id})
                        pipeline.link(src_name, src_cap_filter, src_queue_mux, "mux")
                        sensor_map[i] = SensorInfo(camera_info.camera_csi_sensor_id, f"{i}", src_name)

                elif camera_info.camera_type == "V4L2":
                    if (camera_info.camera_v4l2_dev_node == None):
                        print("camera-v4l2-dev-node required")
                        sys.exit()
                    else:
                        device = f"/dev/video{camera_info.camera_v4l2_dev_node}"
                        pipeline.add("v4l2src", src_name, {"device": device})

                        src_cap_filter1 = f"{src_name}_cap_filter1"
                        capfilterCaps1 = f"video/x-raw, width={camera_info.camera_width}, height={camera_info.camera_height}, framerate={camera_info.camera_fps_n}/{camera_info.camera_fps_d}"
                        pipeline.add("capsfilter", src_cap_filter1, {"caps": capfilterCaps1})

                        nvvidconv2 = f"{src_name}_nvvidconv2"

                        if platform.processor() == "aarch64" and camera_info.nvvideoconvert_copy_hw:
                            pipeline.add("nvvideoconvert", nvvidconv2, {"gpu-id": camera_info.gpu_id if camera_info.gpu_id else 0,
                                                                        "nvbuf-memory-type": camera_info.nvbuf_mem_type if camera_info.nvbuf_mem_type else 0,
                                                                        "copy-hw": camera_info.nvvideoconvert_copy_hw})

                        else:
                            pipeline.add("nvvideoconvert", nvvidconv2, {"gpu-id": camera_info.gpu_id if camera_info.gpu_id else 0,
                                                                        "nvbuf-memory-type": camera_info.nvbuf_mem_type if camera_info.nvbuf_mem_type else 0})
                        pipeline.link(src_name, src_cap_filter1)

                        if platform.processor() != "aarch64":
                            nvvidconv1 = f"{src_name}_nvvidconv1"
                            pipeline.add("videoconvert", nvvidconv1)
                            pipeline.link(src_cap_filter1, nvvidconv1, nvvidconv2)
                        else:
                            pipeline.link(src_cap_filter1, nvvidconv2)

                        pipeline.link(nvvidconv2, src_cap_filter, src_queue_mux, "mux")
                        sensor_map[i] = SensorInfo(device, f"{i}", src_name)
                else:
                    print("Invalid camera-type value")
                    sys.exit()

            pipeline.add("queue", "queue_pgie")
            pipeline.link("mux", "queue_pgie", PGIE)

        else:
            streammux_properties = {
                "batch-size": len(source_config.sensor_list),
                "width": pipeline[TILER].get("width"),
                "height": pipeline[TILER].get("height")
            }
            pipeline.add("nvstreammux", "mux", streammux_properties)
            if "gpu-id" in source_config.source_properties:
                pipeline["mux"].set("gpu-id", source_config.source_properties["gpu-id"])
            for i, sensor_info in enumerate(source_config.sensor_list):
                source_properties["uri"] = sensor_info.uri
                src_name = f"src_{i}"
                pipeline.add(source_config.source_type, src_name, source_properties).link((src_name, "mux"), ("vsrc_%u", ""))
                sensor_map[i] = sensor_info
            pipeline.link("mux", PGIE)
            sr_controller = CommonFactory.create("smart_recording_action", "sr_controller")
            reference_holders.append(sr_controller)

            if isinstance(sr_controller, signal.Emitter):
                sr_controller_properties = {
                        "proto-lib": pipeline[MSGBROKER].get("proto-lib"),
                        "conn-str": pipeline[MSGBROKER].get("conn-str"),
                        "msgconv-config-file": pipeline[MSGCONV].get("config"),
                        "proto-config-file": pipeline[MSGBROKER].get("config"),
                        "topic-list": "test5-sr"
                }
                sr_controller.set(sr_controller_properties)
                for i, _ in enumerate(source_config.sensor_list):
                    source_name = f"src_{i}"
                    sr_controller.attach("start-sr", pipeline[source_name])
                    sr_controller.attach("stop-sr", pipeline[source_name])
                    pipeline.attach(source_name, "smart_recording_signal", "sr", "sr-done")
            else:
                print("Warning: failed to create smart recording controller")

        pipeline.attach(CHECKPOINT, Probe("message_generator", EventMessageGenerator(sensor_map, labels)))

        # OSD mouse event handling
        osd_event_handler = osd.EventHandler(osd=pipeline[OSD], tiler=pipeline[TILER], renderer=pipeline[SINK])
        reference_holders.append(osd_event_handler)
        if platform.processor() != "aarch64":
            osd_event_handler.start()

        # fps measurement
        perf_monitor = utils.PerfMonitor(
            batch_size=len(source_config.camera_list) if use_camerabin else len(source_config.sensor_list),
            interval=args.perf_measurement_interval_sec,
            source_type=source_config.source_type,
            show_name= not args.hide_stream_name)
        perf_monitor.apply(pipeline[TILER], "sink")
        reference_holders.append(perf_monitor)

        # model engine file monitor
        model_engine_monitor = utils.EngineFileMonitor(pipeline[PGIE], pipeline[PGIE].get("model-engine-file"))
        reference_holders.append(model_engine_monitor)
        pipeline.prepare(lambda m, o=osd_event_handler, s=sensor_map, p=perf_monitor, mem=model_engine_monitor: on_message(m, o, s, p, mem))
        pipeline_id += 1
        if platform.processor() == "aarch64":
            osd_event_handler.start()

    for pipeline in pipelines:
        pipeline.activate()

    for pipeline in pipelines:
        pipeline.wait()

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Deepstream Test5')
    parser.add_argument('-c', '--pipeline-config', nargs='+', required=True, help='YAML configuration file for a pipeline')
    parser.add_argument('-s', '--source-config', nargs='+', required=True, help='YAML configuration file for sources of a pipeline')
    parser.add_argument('-l', '--label-file', nargs='*', help='Label file for custom models')
    parser.add_argument('--perf-measurement-interval-sec', type=int, default=1, help='Interval of performance measurement in seconds')
    parser.add_argument('--hide-stream-name', action='store_true', default=False, help="Don't show stream name in fps measurement data")
    args = parser.parse_args()
    
    # pipeline.wait() in the main function is a blocking call due to which the KeyboardInterrupt may not be processed immediately.
    # we use Process from multiprocessing which runs the main function in a different process and processes KeyboardInterrupt immediately.
    process = Process(target=main,  args=(args,))
    try:
        process.start()
        process.join()
    except KeyboardInterrupt:
        print("\nCtrl+C detected. Terminating process...")
        process.terminate()